File: C++/Hard/WrongAnswer/shortest-path-visiting-all-nodes_8.cpp
class Solution {
public:
    int shortestPathLength(vector<vector<int>>& graph) {
        
        int n = graph.size();
        vector<vector<int>> dp(1<<n, vector<int>(n, 1e9));
        for (int i = 0; i < n; i++) {
            dp[1 << i][i] = 0;
        }
        for (int i = 0; i < (1 << n); i++) {
            for (int j = 0; j < n; j++) {
                for (int k : graph[j]) {
                    dp[i | (1 << k)][k] = min(dp[i | (1 << k)][k], dp[i][j] + 1);
                }
            }
        }
        int ans = 1e9;
        for (int i = 0; i < n; i++) {
            ans = min(ans, dp[(1 << n) - 1][i]);
        }
        return ans;
    }
};

File: Java/Hard/Accepted/shortest-path-visiting-all-nodes_8.java
class Solution {
    public int shortestPathLength(int[][] graph) {
        
        int n = graph.length;
        int target = (1 << n) - 1;
        Queue<int[]> q = new LinkedList<>();
        Set<String> visited = new HashSet<>();
        for (int i = 0; i < n; i++) {
            q.offer(new int[]{i, 1 << i, 0});
            visited.add(i + "-" + (1 << i));
        }
        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int u = cur[0], mask = cur[1], dist = cur[2];
            if (mask == target) {
                return dist;
            }
            for (int v : graph[u]) {
                int maskV = mask | (1 << v);
                if (!visited.contains(v + "-" + maskV)) {
                    q.offer(new int[]{v, maskV, dist + 1});
                    visited.add(v + "-" + maskV);
                }
            }
        }
        return -1;
    }
}